composer/callbacks/eval_output_logging_callback.py

```diff
@@ -0,0 +1,68 @@
+# Copyright 2022 MosaicML Composer authors
+# SPDX-License-Identifier: Apache-2.0
+
+"""Log model outputs and expected outputs during ICL evaluation."""
+
+from copy import deepcopy
+
+import torch
+
+from composer.core import Callback, State
+from composer.loggers import Logger
+
+
+class EvalOutputLogging(Callback):
+    """Logs eval outputs for each sample of each ICL evaluation dataset.
+
+    ICL metrics are required to support caching the model's responses including information on whether model was correct.
+    Metrics are responsible for returning the results of individual datapoints in a dictionary of lists.
+    The callback will log the metric name, the depadded and detokenized input, any data stored in state.metric_outputs, and
+    any keys from the batch pased into `batch_keys_to_log`. It will do so after every eval batch.
+    """
+
+    def eval_after_all(self, state: State) -> None:
+        state.metric_outputs = {}
+
+    def eval_batch_end(self, state: State, logger: Logger) -> None:
+        assert state.outputs is not None
+        assert state.metric_outputs is not None
+
+        logging_dict = deepcopy(state.metric_outputs)
+        if state.batch['mode'] == 'generate':
+            # Outputs are already detokenized
+            logging_dict['outputs'] = state.outputs
+        logging_dict['metric_name'] = [state.metric_outputs['metric_name'] for _ in range(0, len(state.outputs))]
```

Comment:
I add the name of the metric once here: https://github.com/mosaicml/composer/pull/2977/files#diff-e61f9af34de66238d99995ebdc5d2645cea3c7ca7c303ac006239a720c5b620cR546

And this line duplicates it so that there's one entry of the metric for each datapoint in the batch (and thus each row will have a copy of it). I could move this to the update_metric function.